networkx/generators/dynamic_networks.py

```diff
@@ -0,0 +1,98 @@
+"""
+Transport Networks
+"""
+import itertools
+import operator
+
+import networkx as nx
+from networkx.utils import not_implemented_for
+
+__all__ = ("gradient_network",)
+
+
+def _value_getter(value):
+    return value() if callable(value) else value
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+def gradient_network(
+    G, scalar_field_value="value", scalar_field_distance="distance", ascending=True
+):
+    """
+    Returns a Gradient Network graph of an input substrate network graph.
+
+    In network science, a Gradient Network is a directed subnetwork of an undirected
+    "substrate" network where each node has an associated scalar potential and one
+    out-link that points to the node with the largest (or smallest) potential in its
+    neighborhood, defined as the union of itself and its neighbors on the substrate
+    network [2]_.
+
+    Parameters
+    ----------
+    G: NetworkX Graph
+        Substrate Network graph. Does not support directed graphs and multigraphs.
+        The graph nodes should have an attribute as a source for the scalar field value.
+        A node `value` attribute should be either a `Number` (e.g.int, float, etc.) or a
+        `Callable` that returns a Number.
+        The graph edges may have an attribute as a source of a `distance` value.
+        An edge `distance` attribute may be either a `Number` or a `Callable` that
+        returns a Number.
+    scalar_field_value: str | Callable[[], int | float]. Default: 'value'.
+        The attribute name for the source of the scalar field value.
+    scalar_field_distance: str Callable[[], int | float]. Default: 'distance'.
```

Comment:
I don't think these callable types are correct, but I could be wrong. The text here should describe how the function works (maybe move that from the description of G above to these scalar_field input arguments). It is also not clear at first which of these are node attributes and which are edge attributes. Perhaps that could be included in the name somehow?  The node "field" and the edge "field" don't refer to the same field right? That could be confusing, so maybe we don't need `field` in the name...  Maybe `node_value`, and `edge_distance`?

The description should say what the function takes for input and output. Something like:
```
    The attribute name for each node's value as stored in the graph,
     or a function that takes a node as input and returns a number. 
```
It looks like right now it doesn't take any input to the function, but it probably would be good to allow the value to depend on which node is being processed.  You should also pass in a node datadict if you think someone might want to have the values depend on node attributes. That's probably the most general approach.

For edge distances, our other edge weight functions (mostly `shortest_path` and friends) take as input a str (attr name) or a function which takes 3 inputs `(node, nbr, datadict)` and returns a number. See [algorithms/shortest_paths/weighted.py::_weight_function](https://github.com/networkx/networkx/blob/1c5272054f71f9484347f7e4246ae3d5da367f7b/networkx/algorithms/shortest_paths/weighted.py#L41) for a helper function to process that input parameter if you want to use that.

Finally, the code asks for the data named by the strings here, but it won't work if e.g. `scalar_field_value` is a function. `G.nodes(data=scalar_field_value)` won't call a function named `scalar_field_value`.  It will however, try to look up a node attribute whose key is a function. You probably need to process the inputs up front and figure out whether it is an attribute name or a function. The `_weight_function` approach turns any input into a functoin that you can use throughout. But it might be better to compute node-values once at the beginning and store the results in a dict. I'm not sure about edge values....